### DISCUSSION TITLE
Crazyflies loosing poses for swarm sizes greater then 6 drones

### DISCUSSION THREAD

**Discussion by darshit-desai** [USC-ACTLab/crazyswarm#769] [Q&A]
Hi @whoenig,

I have been testing the crazyswarm API in a new lab, The lab setup is given as below:

Tracking: Vicon Vantage V16 Cameras working at default 100 FPS. No. of cameras covering the space 40
Area of the space: 80’ by 60’ space with a minimum 30’ ceiling
The lab is RF shielded ( 90 MHz-18 GHz )

Connection with the computer running Crazyswarm: Connected via ethernet cable with the Vicon Computer directly over static IP with separate IP domain

Computer specs running the Crazyswarm: Razer Lambda X Tensorbook running native Ubuntu 22.04

Drone swarm configuration: 
3 Radios on 3 channels over 80, 90 and 100
2 radios connected via USB splitter and 1 radio directly to the Computer

Issue description: The hoverswarm launch script runs and initializes radio and crazyswarm server connection with the vicon and drones, As soon as I start the figure8_csv.py script the drones start loosing pose. Although they takeoff pretty well, but every drone after takeoff starts loosing pose and lands after some time.

What we have debugged?
1) Vicon, we have recalibrated vicon multiple times and also checked the data delivery latency which is well within the specifications of crazyswarm (After all overheads the data is delivered to the Linux computer running crazyswarm at 70 Hz)
2) Tried scaling down the drones from 18--->12 but that also doesn't work
3) Tried using bigger markers and smoother higher quality markers but that also doesn't seem to resolve the issue

What we think might be wrong?

Last week me and @ewinge1  had a zoom call to debug the same issue and the only bottleneck we found was the communication between radios and drones, Eric specifically pointed out that while rebooting using chooser.py script some of the drones timed out and didn't reboot. This could be one of the bottle necks. Since the lab is RF Shielded the radio interference theory could be ruled out. The only other thing we could think of was something in the server script delaying the communication

Comparative testing of the above theory with Crazyswarm 2: (CSW2)

To test the radio failure theory, we tested the same 12 drone swarm on CSW2, the drones did takeoff and some executed figure 8 while some didn't (which is an entirely separate and open issue on CSW2 repo) but none of the drones reported Pose Lost errors. Eric said that since CSW2 uses Safelink protocol the poses are guaranteed to be delivered even though it takes time, while in case of CSW1 there is a timeout which happens, which basically means the radio gives up on the drone.

Contradictions:
I tested a 24 drone swarm using CSW1 in a smaller lab with 12 cameras and it worked just fine (i.e. the figure8 script), plus that lab is not RF shielded and is infact much worser environment for 2.4G devices to get tested.

If you could give me some pointers about this issue it would be greatly helpful

**Comment by whoenig** [comment]
Thanks for the report including the comparison with Crazyswarm2. A few notes

- If Crazyswarm2 works apart from some drones missing unicast commands, you can send those multiple times. This is, as far as I know, what Eric has been doing. This issue is of course very high on our list, although it might require firmware changes, too.
- For the communication, can you run `comCheck` (part of crazyflie_tools)? This can measure the latency, packet drop rates and similar.
- What tracking mode are you using? Single-marker or multiple markers? libobjecttracker or VICON Tracker?
- When you observe "losing pose", is this related to error messages on the terminal, or the drones showing erratic behavior?

**Reply by Eric-Ewing** [reply]
`If Crazyswarm2 works apart from some drones missing unicast commands, you can send those multiple times.` Can confirm that is what we've done here. Darhsit's communication situation is worse than ours for some reason and that has made debugging a bit harder as I can't recreate the problems locally. The biggest red flag to me is the timeouts when calling reboot or sysoff from chooser.py didn't seem to be tied to specific robots or channels. The fact that the timeouts occur without the server running (and thus without pose packets) made me much more concerned about radios than about tracking. I believe he has also seen broadcast messages being missed by some drones in CS2 and not just unicast messages, which we've never had issues with at Brown.

**Reply by whoenig** [reply]
Broadcasts are automatically repeated (and a user can configure how often), so technically this should be fine for bad connections. But the safelink bug together with poor connections might render CS2 inoperable until this firmware bug gets fixed.

**Reply by Eric-Ewing** [reply]
I think even with 30 repeats (twice what we use), they were still missing some commands. It's just an indication of how much worse the connection is than I have locally.

**Reply by darshit-desai** [reply]
Again it is worth mentioning that I tested the same overkill broadcasts and unicasts configuration in a smaller lab with 24 drones and it had some moderate level of success. I boosted the upload trajectory and start trajectory commands by 12 times. (While others were repeated 3 times)

**Comment by darshit-desai** [comment]
For the questions 
* I am running the swarm server using libobjecttracker with single markers.
* The losing pose is a specific warning which appears as soon as the python script starts running, (This warning doesn't come when the hover_swarm launch file is launched). Although during connecting to the drones (i.e., after pressing enter for the launch file launch command), there are multiple warnings of poor link quality until it is able to connect with all the drones. Below is a snapshot of the same pose lost warnings, (This is basically reflected in real world where the drones after takeoff immediately start to drift)
![Screenshot from 2023-09-22 14-48-37(1)](https://github.com/USC-ACTLab/crazyswarm/assets/36150235/94d3bd8d-4c96-4c46-956e-96b873dd4bc1)

For `comCheck` is there a specific command I need to run after the server is launched. I want to clarify the overall sequence of running this diagnostics, is it `rosrun crazyflie_tools comCheck.py` or something

Yes, I had a moderate amount of success with explicit unicasts, and me and Eric found that there are explicitly some commands which even after doing unicasts are not reaching the drone, those are either the upload_trajectory or start_trajectory commands. (Not able to narrow down beyond that)

**Reply by whoenig** [reply]
It should be `rosrun crazyflie_tools comCheck --uri <URI>` (and no server running in parallel).

If the link quality is very poor, there seems to be a bug in safelink that can cause infinite communication, which might have been what you are seeing with Crazyswarm2. It's  documented in https://github.com/bitcraze/crazyflie-firmware/issues/703, but sadly hasn't seen much progress.

**Comment by darshit-desai** [comment]
> If the link quality is very poor, there seems to be a bug in safelink that can cause infinite communication, which might have been what you are seeing with Crazyswarm2. It's documented in [bitcraze/crazyflie-firmware#703](https://github.com/bitcraze/crazyflie-firmware/issues/703), but sadly hasn't seen much progress.

Actually I highlighted this for CSW1, the link quality errors come with CSW1 launch server (and not with CSW2)

**Reply by whoenig** [reply]
Yes, but in CSW2 this is simply not reported (we have statistics on it, but no warning threshold). It is likely that you have the same poor link quality regardless of the Crazyswarm version.

**Reply by darshit-desai** [reply]
Is there some external factor which also affects this like interference from a nearby 2.4G wifi router?

Also, since the bitcraze issue I saw was around 2 years ago, was there any solution at the end? (Sorry I didn't check the latest update, it's still open)

Any other way this can overridden, since CSW1 was already tested on 45 drones at least, there must be a configuration which works?

**Reply by whoenig** [reply]
Any other 2.4 GHz source can cause issues as well as the "structure" of the environment - perhaps your shielding actually reflects the radio waves. The source of a link quality is very difficult to find, but quantifying it is easy using comCheck.

This issue is related to safelink, so CSW1 is not affected. There is currently no workaround other than slowing down the communication itself (and operating in environments where the link quality is very high).

One way to reproduce those issues (for us or Bitcraze) is to flash a custom Crazyradio2 firmware, because there is a compile-time option to simulate a poor radiolink that randomly drops packets. When we implemented that 2 or 3 years ago, most of the software stacks (including bitcrazes own) didn't work properly anymore.

**Reply by darshit-desai** [reply]
Here's what exactly written on the lab website about shielding,

https://matrix.umd.edu/research/open-air-land-lab

```Radio frequency (RF) attenuated 90 MHz-18 GHz (between -10 dB and -30 dB attenuation, depending on frequency) shielding to allow for GPS and other RF testing```

I'll test out the comCheck and reply back, is there a way to run it for multiple URI's at the same time, the command example you gave earlier seemed to be of only one URI, the reason I ask this because, with up to 1-9 drones the latency issue doesn't come, it's at >10 drones when this issues start

**Reply by whoenig** [reply]
You would need to run it in different terminals. GNU parallel or a bash script can automate this.

I am no RF expert, but once we have concrete numbers, we could also check back with Tobias or Arnaud from Bitcraze. It's possible that a simple radio settings change can improve the situation.

**Reply by darshit-desai** [reply]
Ok I will make up a bash script and run in on 12, 18 and 24 drone swarm and post the results. Are the results by default exported to a txt file or do I need to copy the terminal outputs and paste it here?

**Reply by whoenig** [reply]
It's a stdout output. Best to pipe it to a separate file per drone via >>.

**Reply by darshit-desai** [reply]
Hi @whoenig @ewinge1 

I have done the test as you said last to last week, I made a bash file to run the comm check cpp file 10-20 times, and extracted the stdout output in text files for each of the drone, It couldn't be run in parallel since it showed me libusb errors so mostly this was done using one drone at a time.

I have also used a python script later to plot the relevant data points of the drones, the results are in the drive link below along with the raw txt files and the bash files I used to run the ros cpp script.

Note for some cases where even the script timed out, the values of that data point are set to -10.0, which indicates no data was collected.

There are two folders containing two sets of results, One is from the new lab, the folder is called USMSM Testing and one is where I tested 24 drones successfully, the folder is called UMD Testing.

The below is the link to the files I have uploaded on google drive:

https://drive.google.com/drive/folders/1q-Lz7YaW1PZRHImvUxSIJsM_YmXzJeBx?usp=sharing

The below is the folder structure:

```BASH
.
├── read_and_plot_UMD.py
├── read_and_plot_USMSM.py
├── Results
│   ├── UMD Testing
│   │   ├── with 15 drones
│   │   └── with24drones
│   └── USMSM testing
│       ├── with 15 drones
│       ├── with 24 drones
│       ├── with 24 drones 2nd time
│       └── With 9 drones
├── UMD Testing
│   ├── with 15 drones
│   │   ├── commcheckcli.txt
│   │   ├── comm_check.sh
│   │   ├── drone10_commcheck.txt
│   │   ├── drone11_commcheck.txt
│   │   ├── drone12_commcheck.txt
│   │   ├── drone13_commcheck.txt
│   │   ├── drone14_commcheck.txt
│   │   ├── drone15_commcheck.txt
│   │   ├── drone1_commcheck.txt
│   │   ├── drone2_commcheck.txt
│   │   ├── drone3_commcheck.txt
│   │   ├── drone4_commcheck.txt
│   │   ├── drone5_commcheck.txt
│   │   ├── drone6_commcheck.txt
│   │   ├── drone7_commcheck.txt
│   │   ├── drone8_commcheck.txt
│   │   └── drone9_commcheck.txt
│   └── with24drones
│       ├── commcheckcli.txt
│       ├── comm_check.sh
│       ├── drone10_commcheck.txt
│       ├── drone11_commcheck.txt
│       ├── drone12_commcheck.txt
│       ├── drone13_commcheck.txt
│       ├── drone14_commcheck.txt
│       ├── drone15_commcheck.txt
│       ├── drone16_commcheck.txt
│       ├── drone17_commcheck.txt
│       ├── drone18_commcheck.txt
│       ├── drone19_commcheck.txt
│       ├── drone1_commcheck.txt
│       ├── drone20_commcheck.txt
│       ├── drone21_commcheck.txt
│       ├── drone22_commcheck.txt
│       ├── drone23_commcheck.txt
│       ├── drone24_commcheck.txt
│       ├── drone2_commcheck.txt
│       ├── drone3_commcheck.txt
│       ├── drone4_commcheck.txt
│       ├── drone5_commcheck.txt
│       ├── drone6_commcheck.txt
│       ├── drone7_commcheck.txt
│       ├── drone8_commcheck.txt
│       └── drone9_commcheck.txt
└── USMSM testing
    ├── with 15 drones
    │   ├── comm_check.sh
    │   ├── drone10_commcheck.txt
    │   ├── drone11_commcheck.txt
    │   ├── drone12_commcheck.txt
    │   ├── drone13_commcheck.txt
    │   ├── drone14_commcheck.txt
    │   ├── drone15_commcheck.txt
    │   ├── drone1_commcheck.txt
    │   ├── drone2_commcheck.txt
    │   ├── drone3_commcheck.txt
    │   ├── drone4_commcheck.txt
    │   ├── drone5_commcheck.txt
    │   ├── drone6_commcheck.txt
    │   ├── drone7_commcheck.txt
    │   ├── drone8_commcheck.txt
    │   └── drone9_commcheck.txt
    ├── with 24 drones
    │   ├── comm_check_logs.txt
    │   ├── comm_check.sh
    │   ├── drone10_commcheck.txt
    │   ├── drone11_commcheck.txt
    │   ├── drone12_commcheck.txt
    │   ├── drone13_commcheck.txt
    │   ├── drone14_commcheck.txt
    │   ├── drone15_commcheck.txt
    │   ├── drone16_commcheck.txt
    │   ├── drone17_commcheck.txt
    │   ├── drone18_commcheck.txt
    │   ├── drone19_commcheck.txt
    │   ├── drone1_commcheck.txt
    │   ├── drone20_commcheck.txt
    │   ├── drone21_commcheck.txt
    │   ├── drone22_commcheck.txt
    │   ├── drone23_commcheck.txt
    │   ├── drone24_commcheck.txt
    │   ├── drone2_commcheck.txt
    │   ├── drone3_commcheck.txt
    │   ├── drone4_commcheck.txt
    │   ├── drone5_commcheck.txt
    │   ├── drone6_commcheck.txt
    │   ├── drone7_commcheck.txt
    │   ├── drone8_commcheck.txt
    │   └── drone9_commcheck.txt
    ├── with 24 drones 2nd time
    │   ├── commcheckcli.txt
    │   ├── comm_check.sh
    │   ├── drone10_commcheck.txt
    │   ├── drone11_commcheck.txt
    │   ├── drone12_commcheck.txt
    │   ├── drone13_commcheck.txt
    │   ├── drone14_commcheck.txt
    │   ├── drone15_commcheck.txt
    │   ├── drone16_commcheck.txt
    │   ├── drone17_commcheck.txt
    │   ├── drone18_commcheck.txt
    │   ├── drone19_commcheck.txt
    │   ├── drone1_commcheck.txt
    │   ├── drone20_commcheck.txt
    │   ├── drone21_commcheck.txt
    │   ├── drone22_commcheck.txt
    │   ├── drone23_commcheck.txt
    │   ├── drone24_commcheck.txt
    │   ├── drone2_commcheck.txt
    │   ├── drone3_commcheck.txt
    │   ├── drone4_commcheck.txt
    │   ├── drone5_commcheck.txt
    │   ├── drone6_commcheck.txt
    │   ├── drone7_commcheck.txt
    │   ├── drone8_commcheck.txt
    │   └── drone9_commcheck.txt
    └── With 9 drones
        ├── comm_check.sh
        ├── drone1_commcheck.txt
        ├── drone2_commcheck.txt
        ├── drone3_commcheck.txt
        ├── drone4_commcheck.txt
        ├── drone5_commcheck.txt
        ├── drone6_commcheck.txt
        ├── drone7_commcheck.txt
        ├── drone8_commcheck.txt
        └── drone9_commcheck.txt

17 directories, 123 files

```

**Reply by whoenig** [reply]
Very strange, so in the old lab everything looks normal. In the new lab everything looks normal with 9 CFs and it starts looking weird with 15. The used radios and channels are identical between 9 and 15 (and many of the drones that behave differently), which likely means that their physical placement causes the difference. (I have seen such behavior before, but not with such extreme drop rates in the radio performance.)

If you have Crazyflie 2.1s, they do have an external antenna port. It seems like bitcraze doesn't sell antennas, but I remember that they told me once that one can use the one that comes with the Bolt (if you have one). Otherwise you can contact them and check if they can recommend a matching product. They told me that this significantly improves radio link quality over the onboard chip antenna.

I don't think there is much we can do on the Crazyswarm1 side with such poor communication (sometimes the link quality is 30% or lower). For Crazyswarm2, we are working on a way to collect this kind of data automatically during a normal execution, so we can at least warn users who experience problems. Nevertheless, a hardware change or different settings (channels) will be required to achieve satisfactory flight performance regardless.

**Reply by darshit-desai** [reply]
About the first part, the arrangement of the drones is quite sparse, with each drone having at least 1 metres gap between each other, the radios are usually placed about 8 metres away from the farthest drone and 3 metres away from the nearest drone.

What would you suggest in changing the channels? Is it spacing them more? Currently it's 80, 90 and 100 between the 3 radios. Would also increasing the number of radios help in this scenario? , Combined with channel allocation changes which would look something like 4 radios communicating on channels 40,60,80 and 100?

Also notice in the above results all of the tests were in unicast mode since libusb blocks running the script in multicast

**Reply by whoenig** [reply]
Your test was executed serially and showed the problem, so changing the number of radios will not have an effect.

It's possible that some channels behave better, but that has nothing to do with the "spacing" between them (the frequencies don't overlap already). You can try finding better channels using [https://github.com/bitcraze/crazyflie-firmware/ examples/radio/radio-test.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/radio/radio-test.py).

**Reply by darshit-desai** [reply]
Could it be the environment of the lab which is responsible for this too? Although the place where the results are poor has much better RF shielding

**Reply by whoenig** [reply]
I strongly assume it has something to do with the environment, since the same (?) drones and settings worked fine in the other lab. I am not an RF expert, but perhaps the shielding prevents multipath reflections that are "useful" for those small chip antennas.

**Reply by darshit-desai** [reply]
Was the purpose of getting the above diagnostics data to ask Bitcraze RF engineers for a solution?

**Reply by whoenig** [reply]
It was mostly to rule out that the problems you are seeing are related to Crazyswarm (1 or 2), but rather an issue with the specific environment. While we can do a few more things in Crazyswarm2 to support poor communication better, some of your message outages were so significant that other changes (such as an external antenna) are likely to be needed.

**Reply by darshit-desai** [reply]
Can we contact them anyways? I think that this is not the last time anyone is testing it in a RF shielded environment and there must be solutions like you suggested, mounting antennas on the drones, similar might be available for the radios too, maybe a different type of antenna or something. And if it is true that the radios rely on reflected waves (and the lab walls are absorbing those) there might be stopgap solutions like mounting radios on the drones.

**Reply by whoenig** [reply]
Sure, just post on https://github.com/orgs/bitcraze/discussions, and also tag me and include a link to the discussion here.

**Reply by darshit-desai** [reply]
Thanks @whoenig just mentioned this issue in this thread https://github.com/orgs/bitcraze/discussions/1016#discussion-5806395 on bitcraze discussions
